### PR TITLE
feat: Add vector projection and inversion

### PR DESCRIPTION
**Pull request by spydon**
# Description

Adding `projection`, `ìnvert` and `inverted` to the vector extension.

## Checklist

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [x] I have read the [Contributor Guide] and followed the process outlined for submitting PRs.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples`.

## Breaking Change

- [x] No, this is *not* a breaking change.

## Related Issues

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md
[Conventional Commit]: https://conventionalcommits.org

### CODE REVIEW

File: packages/flame/lib/src/extensions/vector2.dart [hunk 1]
```diff
@@ -80,6 +80,26 @@ extension Vector2Extension on Vector2 {
...
   }
 
+  /// Project this onto [other].
+  ///
+  /// If [out] is specified, it will be used to provide the result.
+  Vector2 projection(Vector2 other, {Vector2? out}) {
+    final dotProduct = dot(other);
+    final result = (out?..setFrom(other)) ?? other.clone();
+    return result..scale(dotProduct / other.length2);
+  }
```
Comment (st-pasha) [inline packages/flame/lib/src/extensions/vector2.dart:90]: What if `other` is zero? If we don't want to allow this case, then at least mention it in the docs.

File: packages/flame/lib/src/extensions/vector2.dart [hunk 2]
```diff
@@ -80,6 +80,26 @@ extension Vector2Extension on Vector2 {
...
+
+  /// Inverts the vector.
+  void invert() {
+    x *= -1;
+    y *= -1;
+  }
+
+  /// Returns the inverse of this vector.
+  Vector2 inverted() {
+    return Vector2.all(-1)..multiply(this);
```
Comment (st-pasha) [inline packages/flame/lib/src/extensions/vector2.dart:100]: Probably simpler to have `Vector2(-x, -y)`

File: packages/flame/test/extensions/vector2_test.dart [hunk 3]
```diff
@@ -208,4 +208,42 @@ void main() {
...
+      expect(result, Vector2(5, 0));
+    });
+
+    test('Project onto vector in other direction', () {
+      final u = Vector2(5, 2);
+      final v = Vector2(-10, 0);
+      final result = u.projection(v);
+      expect(result, Vector2(5, 0));
+    });
+  });
```
Comment (st-pasha) [inline packages/flame/test/extensions/vector2_test.dart:233]: Seems to be no tests for the `out` argument

File: packages/flame/lib/src/extensions/vector2.dart (hunk 1 above)
Comment (spydon) [reply packages/flame/lib/src/extensions/vector2.dart:92]: Added an assert and to the docs.

File: packages/flame/test/extensions/vector2_test.dart (hunk 3 above)
Comment (spydon) [reply packages/flame/test/extensions/vector2_test.dart:250]: Fixed

**Review by st-pasha — APPROVED** [review]

**Review by erickzanardo — APPROVED** [review]

File: packages/flame/lib/src/extensions/vector2.dart [hunk 4]
```diff
@@ -80,6 +80,26 @@ extension Vector2Extension on Vector2 {
     }
   }
 
+  /// Project this onto [other].
+  ///
+  /// [other] needs to have a length > 0;
+  /// If [out] is specified, it will be used to provide the result.
+  Vector2 projection(Vector2 other, {Vector2? out}) {
+    assert(other.length2 > 0, 'other needs to have a length > 0');
```
Comment (wolfenrain) [inline packages/flame/lib/src/extensions/vector2.dart:88]: imho, we should have a test for this assertion as well to ensure it stays

**Review by wolfenrain — APPROVED** [review]

**Pull request merged**